### PR TITLE
nixos/ddclient: make nsupdate protocol usuable

### PR DESCRIPTION
**Pull request by Mic92**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/ddclient.nix [hunk 1]
```diff
@@ -29,8 +29,10 @@ let
   configFile = if (cfg.configFile != null) then cfg.configFile else configFile';
 
   preStart = ''
-    install ${configFile} /run/${RuntimeDirectory}/ddclient.conf
-    ${lib.optionalString (cfg.configFile == null) (if (cfg.passwordFile != null) then ''
+    install --owner ddclient -m600 ${configFile} /run/${RuntimeDirectory}/ddclient.conf
```
Comment (Mic92) [inline nixos/modules/services/networking/ddclient.nix:32]: This mainly avoids warnings but should have not be strictly necessary.

**Comment by SuperSandro2000** [comment]
Can you fix the manual?

**Comment by Mic92** [comment]
fixed.

**Comment by Mic92** [comment]
> Can everyone please stop breaking this module without testing it? This is the exact same problem as [last time](https://github.com/NixOS/nixpkgs/pull/149306).

I tested this module.

> 
> ```
> Jan 13 10:51:47 shanghai 3jifr0sab601i8l2ka07c2c2cak5csg2-ddclient-prestart[948280]: install: invalid user ‘ddclient’
> │ Jan 13 10:51:47 shanghai 3jifr0sab601i8l2ka07c2c2cak5csg2-ddclient-prestart[948284]: sed: can't read /run/ddclient/ddclient.conf: No such file or directory
> ```
> 
> It's a `DynamicUser` service, you can't just chown it like that AFAIK

This is not true. If it cannot find the user than something is wrong with your nss setup. Systemd creates this user before starting `PreStart`. Was nscd by chance restarted in the middle?

**Comment by Mic92** [comment]
In https://github.com/NixOS/nixpkgs/commit/1024571d35ac1d3329e8475c7b5e78c345230dab I fixed some race conditions that broke dynamic user in some cases. What is the full restart log, when you run an upgrade?

**Comment by arcnmx** [comment]
> This is not true. If it cannot find the user than something is wrong with your nss setup.

Thanks for the hint, I'll look into that. I'd been trying to look for documentation on how `DynamicUser` interacts with exec/etc for ages now .-.

> Systemd creates this user before starting PreStart. Was nscd by chance restarted in the middle?

Possibly! It certainly broke a switch :(

> What is the full restart log, when you run an upgrade?

I can try to look into it in a bit but if it indeed is related to nscd updating at the same time, I've already performed the update by now ><

**Comment by Mic92** [comment]
It looks like my nscd fix is no longer in nixpkgs. I re-introduced it in this PR: https://github.com/NixOS/nixpkgs/pull/154928
Can you test it?

**Comment by Mic92** [comment]
> just FYI so that you know: I am the more or less the upstream (core) maintainer of ddclient as of yesterday.

dual-stack support please :)
For this reason I had to use this code instead: https://github.com/Mic92/dotfiles/blob/1348780fa97092731f0b2f483580c8cbd9eff213/nixos/modules/ip-update.nix#L29

**Comment by arcnmx** [comment]
> dual-stack support please :)

wasn't that added some time ago? [I'm pretty sure it's working for me](https://github.com/arcnmx/home/blob/ee8379940aed0575968dac21057b06f22fe69be8/config/profiles/personal/ddclient.nix#L75-L80).

**Comment by Mic92** [comment]
Ok. Configuration was not clear to me at all. I saw something that looked like an RFC. Might be also that the nsupdate backend does not handle this.

**Comment by arcnmx** [comment]
> I saw something that looked like an RFC

https://github.com/ddclient/ddclient/commit/24ba9459491197fe045934a3e7fb54655812cb25 appears to be the result of that and the introduction of proper support, but there hasn't been a tagged release in 2 years, so...

> Might be also that the nsupdate backend does not handle this

Maybe yeah, I'm not sure what support for it looks like for individual backends :<

**Comment by glittershark** [comment]
is there a workaround short of depending on #154928 to get ddclient working again?

**Comment by SuperSandro2000** [comment]
> dual-stack support please :)

Write me a private message and we can figure this out.

> but there hasn't been a tagged release in 2 years, so...

In the meantime I created a fork which I am planning to upstream in the next weeks and then create an RC maybe in a month or so. No promises though.

**Pull request merged**